Adding Drawer controller

kubalani opened the pull request:
### Platforms Impacted
- [X] iOS
- [ ] macOS

### Description of changes

Adding Drawer Controller SwitUI component. Added design tokens for shadow modification.
Currently the drawer only supports slide out cases for left and right presentation style.  

Edit: 

Updated presentations style to override modal screen size. 
![Screen Shot 2020-12-21 at 6 33 11 PM](https://user-images.githubusercontent.com/63682282/102842153-2abe6c00-43bb-11eb-9491-33955146b56a.png)

#### Left Presentation Style
![Screen Shot 2020-12-15 at 7 22 34 PM](https://user-images.githubusercontent.com/63682282/102301183-269dd480-3f0b-11eb-8551-08defe0bd918.png)

#### Right Presentation Style
![Screen Shot 2020-12-15 at 7 22 47 PM](https://user-images.githubusercontent.com/63682282/102301226-39b0a480-3f0b-11eb-80b7-c6eafec9de81.png)


#### Presentation with Dimmed background 
![Screen Shot 2020-12-15 at 7 22 59 PM](https://user-images.githubusercontent.com/63682282/102301249-446b3980-3f0b-11eb-9b7b-d75f022dccba.png)


### Pull request checklist

This PR has considered:
- [ ] Light and Dark appearances
- [ ] VoiceOver and Keyboard Accessibility
- [ ] Internationalization and Right to Left layouts
- [ ] Different resolutions (1x, 2x, 3x)
- [ ] Size classes and window sizes (iPhone vs iPad, notched devices, multitasking, different window sizes, etc)


###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/microsoft/fluentui-apple/pull/344)

harrieshin commented:
why is the drawer not shown on top of the nav bar?

harrieshin commented on ios/FluentUI.Demo/FluentUI.Demo/Demos/DrawerVnextDemoController.swift:
```diff
@@ -0,0 +1,181 @@
+        container.addArrangedSubview(controller.view)
+        self.drawerController = controller
+
+        controller.view.translatesAutoresizingMaskIntoConstraints = false
```
isn't it weird that client app need to do this?

rdeassis replied:
I think the the init() of the wrapper class is the best place for it.

kubalani replied:
This is mostly required to override the default behavior of `UIStackView`. 
Ideally I would expect the client to directly present or transition to `VXTDrawerController` without the need of `DemoController` which is how I think Teams should consume it.

kubalani replied:
I updated this to use `MSFDrawerVnext` extension instead of using view controller API's

harrieshin commented on ios/FluentUI/DrawerVnext/Drawer.swift:
```diff
@@ -0,0 +1,224 @@
+
+public class DrawerTokens: ObservableObject {
+
+    @Published public var shadowColor: Color! = Color.black
```
avoid using "!" in property declarations

harrieshin commented on ios/FluentUI/DrawerVnext/Drawer.swift:
```diff
@@ -0,0 +1,224 @@
+@objc(DrawerState)
+
+public class DrawerState: NSObject, ObservableObject {
+    @objc @Published public var presentationDirection: DrawerDirection = .left
```
all of these properties should be properly documented

rdeassis commented on ios/FluentUI/DrawerVnext/Drawer.swift:
```diff
@@ -0,0 +1,224 @@
+@objc(VXTDrawerController)
+
+/// UIKit wrapper that exposes the SwiftUI Button implementation
+open class VXTDrawerController: NSObject {
```
How about naming this like this?

@objc(**MSFDrawerVnext**)
open class **DrawerVnext**: NSObject

rdeassis commented on ios/FluentUI/DrawerVnext/Drawer.swift:
```diff
@@ -0,0 +1,224 @@
+            }
+            .onEnded { _ in
+                if let draggedOffsetWidth = draggedOffsetWidth {
+                    if abs(draggedOffsetWidth) < 270 / 4 {
```
Where is 270 and 4 coming from?
Shouldn't this be relative to the width?

kubalani replied:
Thanks for catching this, I have updated this to use geometry reader instead

rdeassis commented on ios/FluentUI/DrawerVnext/Drawer.swift:
```diff
@@ -0,0 +1,224 @@
+                    Spacer()
+                }
+
+            }.onTapGesture {
```
nit: put the modifiers on the next line?

kubalani commented on ios/FluentUI/DrawerVnext/Drawer.swift:
```diff
@@ -0,0 +1,224 @@
+
+public class DrawerTokens: ObservableObject {
+
+    @Published public var shadowColor: Color! = Color.black
+    @Published public var shadowOpacity: Double! = 0.24
+    @Published public var shadowBlur: CGFloat! = 28
+    @Published public var shadowDepth: [CGFloat]! =  [0, 14.0] // x,y axis
```
These defaults are coming from the figma file. I will update them to optionals

kubalani commented on tools/sgen/input/FluentUIStyle.yml:
```diff
@@ -147,6 +147,13 @@ Opacity:
+    - light12: 28pt
+
+    Depth:
+    - shadow28: [ 0pt, 14pt ]
```
I did this initially seems like Sgen doesn't like nesting beyond 3 levels, it crashed on me.

rdeassis replied:
You can do it like this:

Shadow:
  shadow28: [ blur: 28pt, color: $Colors.Neutral.black, opacity: $Opacity.opacity24, x: 0pt, y: 14pt ]

AP_DrawerTokens:
  shadowColor: $Shadow.shadow28.color
  shadowOpacity: $Shadow.shadow28.opacity
  shadowBlur:   $Shadow.shadow28.blur
  shadowX: $Shadow.shadow28.x
  shadowY: $Shadow.shadow28.y

kubalani replied:
I have reached out to victor to double check on this. I have created a new issue to track this item. 
https://github.com/microsoft/fluentui-apple/issues/358

kubalani commented on ios/FluentUI/DrawerVnext/Drawer.swift:
```diff
@@ -0,0 +1,224 @@
+        super.init()
+    }
+
+    struct DrawerContentView: UIViewRepresentable {
```
I am not sure if I entirely understand your proposal, I agree with a convenience UIView API but that would be in the wrapper class as well

kubalani replied:
I have updated the API to remove presentation controller. The ShimView is still required to bridge the gap with modal presentation style. I also tried content view with `UIViewRepresentable` but it seems like in that case the controller doesn't receive touch events. 

I have also updated the API to match interface with FluentUI `DrawerController`. I think this is the minimal hosting code to feature complete teams requirement. We can discuss offline if there is a need for additional API

kubalani commented:
> why is the drawer not shown on top of the nav bar?

The API of DemoController is now matched to show full screen.

rdeassis commented on ios/FluentUI/DrawerVnext/Drawer.swift:
```diff
@@ -0,0 +1,266 @@
+        return 0
+    }
+
+    public func animateTransition(using transitionContext: UIViewControllerContextTransitioning) {
```
I found a bug where if you drag from the right it will display the drawer from the left side anyways:

https://user-images.githubusercontent.com/68076145/102934669-71fb3a00-4459-11eb-8691-1c2660cddf17.mov

rdeassis replied:
The other bug also shown in this video is the one you mentioned before about displaying it from the left and then from the right.

kubalani replied:
The left to right pan gesture is correct, as far the animation when switching side is concerned would be addressed as part of this [issue](https://github.com/microsoft/fluentui-apple/issues/357)

rdeassis commented on ios/FluentUI/DrawerVnext/Drawer.swift:
```diff
@@ -0,0 +1,266 @@
+                    Spacer()
+                }
+                content
+                    .frame(width: reader.size.width * percentWidthOfContent)
```
The current drawer does not change its width in horizontal orientation.
This seems to be the right behavior for the left nav at least.

![drawer_horizontal](https://user-images.githubusercontent.com/68076145/102935099-4f1d5580-445a-11eb-86b9-b1afa3929cfc.png)
![drawer_vnext_horizontal](https://user-images.githubusercontent.com/68076145/102935103-517faf80-445a-11eb-8b1c-3697c8235fb1.png)

kubalani replied:
Thanks for catching this, I have updated to match the default functionality. [I've added a small hack to override the default behavior] 
I believe the approach should be to either use `preferredContentSize` from content view controller or stretch to full screen because if any app needs the functionality other way around then currently there is no way to do so. 

We can revisit this when adding vertical support for drawer controller.

rdeassis commented on ios/FluentUI/DrawerVnext/Drawer.swift:
```diff
@@ -0,0 +1,266 @@
+                    Spacer()
+                }
+            }
+            .background(state.isExpanded ? backgroundLayerColor.opacity(backgroundLayerOpacity) : Color.clear)
```
You can create a GitHub issue to address this after your first merge:
If you tap out of the drawer (dimmed background or just out of its contents) you should be able to dismiss it:


https://user-images.githubusercontent.com/68076145/102935346-d4086f00-445a-11eb-8a45-e641139c81d9.mov

rdeassis replied:
Thanks for fixing this!

rdeassis commented on ios/FluentUI/DrawerVnext/Drawer.swift:
```diff
@@ -0,0 +1,266 @@
+        }
+    }
+
+    private func dragGesture(snapWidth: CGFloat) -> some Gesture {
```
Another issue you can file as a GitHub issue to be addressed after your first merge:

The current drawer presents the contents as you drag from the side instead of detecting the drag and presenting it entirely:


https://user-images.githubusercontent.com/68076145/102935613-585af200-445b-11eb-8a9c-99b48f8a66a7.mov

kubalani replied:
I have created a new issue to track this. 
https://github.com/microsoft/fluentui-apple/issues/357

rdeassis commented on ios/FluentUI/DrawerVnext/Drawer.swift:
```diff
@@ -0,0 +1,266 @@
+        if isPresentingDrawer {
+            transitionContext.containerView.addSubview(drawerView)
+            drawerView.frame = UIScreen.main.bounds
+            // Added a static interval as currently swiftui doesn't have an animation completion callback
```
Do you think that the withAnimation function could be useful to avoid this?

https://developer.apple.com/documentation/swiftui/withanimation(_:_:)

rdeassis replied:
By nesting animation modifiers I believe you can eventually disable the animation of repositioning the drawer while animating just its presentation and avoiding the bug on the video below:


https://user-images.githubusercontent.com/68076145/103310566-c314ac00-49cc-11eb-9a00-b1c2af5c6643.mov

kubalani replied:
This should be now working as expected

rdeassis commented on ios/FluentUI/DrawerVnext/Drawer.swift:
```diff
@@ -0,0 +1,266 @@
+}
+
+@objc(DrawerStateDelegate)
+
```
nit: remove the extra line (27) here and in other types.

rdeassis commented on ios/FluentUI.Demo/FluentUI.Demo/Demos/DrawerVnextDemoController.swift:
```diff
@@ -0,0 +1,145 @@
+
+        var views = [UIView]()
+        views.append(createButton(title: "Dismiss", action: #selector(dismissButtonTapped)))
+        views.append(createButton(title: "Dismiss (no animation)", action: #selector(dismissNotAnimatedButtonTapped)))
```
FYI: When you pull the latest changes to your branch, you'll need to change the usage of the button to use the ButtonVnext closure instead of the selector.

rdeassis commented on ios/FluentUI.Demo/FluentUI.Demo/Demos/DrawerVnextDemoController.swift:
```diff
@@ -0,0 +1,145 @@
+            return
+        }
+
+        let leadingEdge: UIRectEdge = view.effectiveUserInterfaceLayoutDirection == .leftToRight ? .left : .right
```
hmmmm... I think the client app shouldn't need to worry about this.
Isn't there a way to just attach the drawerController to a specific view/viewcontroller and let it read the gestures and take the appropriate actions?

kubalani replied:
I have created a new issue to track this 

https://github.com/microsoft/fluentui-apple/issues/356

rdeassis replied:
I have added a comment referencing the code of the current DrawerController to clarify that this logic should live in the control instead of the client app.

rdeassis commented on ios/FluentUI.Demo/FluentUI.Demo/Demos/DrawerVnextDemoController.swift:
```diff
@@ -0,0 +1,145 @@
+        if let drawerController = drawerController {
+            drawerController.drawerState.backgroundDimmed = true
+            drawerController.drawerState.presentationDirection = .right
+            present(drawerController, animated: true, completion: nil)
```
How would a SwiftUI application use/call the drawer in order to be displayed?

kubalani replied:
I have added SwiftUI drawer preview code for an application using SwiftUI

kubalani commented on tools/sgen/input/FluentUIStyle.yml:
```diff
@@ -211,3 +218,8 @@ SecondaryButtonTokens extends MSFButtonTokens:
 GhostButtonTokens extends MSFButtonTokens:
   borderColor: [ rest: $Colors.Neutral.clear, hover: $Colors.Neutral.clear, pressed: $Colors.Neutral.clear, selected: $Colors.Brand.shade10, disabled: $Colors.Neutral.clear ]
 
+AP_DrawerTokens:
```
The drawer inherits the background color from content view controller. The default color is `.clear`. Overriding it won't have any effect when content color is set.

rdeassis replied:
Makes sense!

rdeassis commented on ios/FluentUI/DrawerVnext/Drawer.swift:
```diff
@@ -0,0 +1,331 @@
+                        Image(systemName: "sidebar.left")
+                    })).background(Color.blue)
+            }
+            drawer
```
drawer.edgesIgnoringSafeArea(.all)

(so we can display the drawer in the whole screen?)

kubalani replied:
This is already part of `DrawerShimView`.

rdeassis commented on ios/FluentUI/DrawerVnext/Drawer.swift:
```diff
@@ -0,0 +1,331 @@
+}
+
+/**
+ `Drawer` is used to present a content that is currently in hideout mode when collapsed and when expanded is used to present content view partially either on the left or right.
```
nit: Xcode style comment

sophialee0416 replied:
^this means using `///` instead of the other commenting methods for clarification

rdeassis reviewed (commented):
I think the issue on the video below is the main blocker.
Figuring out the animation comment might solve that comment along with this bug:


https://user-images.githubusercontent.com/68076145/103310694-10911900-49cd-11eb-891d-dc8d79172f17.mov

sophialee0416 commented on ios/FluentUI/DrawerVnext/Drawer.swift:
```diff
@@ -0,0 +1,331 @@
+    private let percentWidthOfContent: CGFloat = 0.9
+    private let percentSnapWidthOfScreen: CGFloat = 0.225
+
+    // content view on top of
```
Is this comment incomplete?

kubalani replied:
Thanks for catching this, I have updated it

kubalani commented on ios/FluentUI/DrawerVnext/Drawer.swift:
```diff
@@ -0,0 +1,331 @@
+    }
+}
+
+#if DEBUG
```
For the meantime, I have added a preview with example for SwiftUI in a separate file

kubalani commented on ios/FluentUI/DrawerVnext/Drawer.swift:
```diff
@@ -0,0 +1,331 @@
+
+extension DrawerVnext: DrawerStateDelegate {
+    // two way binding if state is changed
+    func drawerDidCollapsed() {
```
this method no longer exists

sophialee0416 commented on ios/FluentUI/DrawerVnext/Drawer+Drag.swift:
```diff
@@ -0,0 +1,31 @@
+    }
+
+    private func gestureSnapWidthRatio() -> CGFloat {
+        return 0.225
```
Should this be tokenized?

kubalani replied:
@sophialee0416 tokens are only applied to appearances and not behavioral attributes

@rdeassis  we can't add const static property in Drawer extension because of it's generic type. 
I am also not a big fan of redundant documentation. 

this is from FluentUI [drawer](https://github.com/microsoft/fluentui-apple/blob/bb4d1ac255b858755f25ea427ad2305a66c90b0f/ios/FluentUI/Drawer/DrawerController.swift#L262) 
```swift
// If the drawer is already loaded, update the resizing handle ...
 if isViewLoaded {
       updateResizingHandleView()
}
```

Redundant documentation creates a lot of problems than in solves. documentation also goes out of sync more often that code does.

I am like to add comments when they are [necessary](https://blog.cleancoder.com/uncle-bob/2017/02/23/NecessaryComments.html). 

I intentionally create small method where it's name eliminates the need for documentation and I believe `gestureSnapWidthRatio` in Drag extension has sufficient context to interpret the meaning of the value. I am open to other suggestions to improve nomenclature or restructure code to eliminate the need of documentation.

sophialee0416 commented on ios/FluentUI/DrawerVnext/Drawer+Modifiers.swift:
```diff
@@ -0,0 +1,21 @@
+    /// Custom modifier for adding a callback placeholder when drawer's state is changed
+    /// - Parameter `didChangedState`: closure executed with drawer is expanded or collapsed
+    /// - Returns: `Drawer`
+    func didChangedState(_ didChangedState: @escaping DrawerStateChangedCompletionBlock) -> Drawer {
```
nit: didChangeState

sophialee0416 commented on ios/FluentUI/DrawerVnext/Drawer.swift:
```diff
@@ -0,0 +1,170 @@
+    }
+}
+
+// MARK: - Previews
```
Are previews something we want to check in? I think they're a good idea if someone wants to see the live previews within Xcode, but they could also test the demo app.. Thoughts?

rdeassis replied:
I agree we should question ourselves about it and it's why I created #360.

My first gut reaction is that this should be part of the demo app code to prevent increasing the library size for consumers out of the cocoapods world.

kubalani replied:
I would like to keep previews for testing till we have a solution.. We can move them to appropriate place when we address #360

sophialee0416 commented on ios/FluentUI/DrawerVnext/SlideOverPanel.swift:
```diff
@@ -0,0 +1,174 @@
+
+// MARK: - Preview
+
+struct SlideOverPanelLeft_Previews: PreviewProvider {
```
Re: preview comment from before

kubalani replied:
I will address them as part of #360

rdeassis commented on ios/FluentUI/DrawerVnext/Drawer+Modifiers.swift:
```diff
@@ -0,0 +1,21 @@
+
+import SwiftUI
+
+public extension Drawer {
```
Same thing here regarding the need for a separate file and an extension: I think this should go to the Drawer file as a func and the extension should be deleted unless we have more modifiers to justify it.

kubalani replied:
I have used extension when there is a logical grouping of functionalities, modifiers are going to be extended as we add vertical drawer. The extensions in Drawer are split based on UIKit/SwiftUI native interfaces. As the drawer will grow there would be necessity to clarify the distinction. A client might end up using the other interface if there is not sufficient clear distinction between them. we can discuss this more offline if required

rdeassis replied:
I can see the value of creating separate files for modifiers related to a certain functionality.
My understanding of your idea in this case would be something like Drawer+Delegate.swift or Drawer+State.swift.
Drawer+Modifiers.swift seems like a file that will contain all modifiers for it irrespective of their functionality group.
We can indeed discuss offline.

kubalani replied:
I will remove the drawer modifier for now as it contains only single API. We can discuss this later as we extend the functionality and more API's in drawer

rdeassis commented on ios/FluentUI/DrawerVnext/Drawer.swift:
```diff
@@ -0,0 +1,170 @@
+    /// Set to `false` to restore it to the normal size.
+    @objc @Published public var isExpanded: Bool = false {
+        didSet {
+            onStateChange?(isExpanded)
```
Do we need to pass isExpanded as a parameter? The state object property could just be queried instead.

rdeassis commented on ios/FluentUI/DrawerVnext/Drawer.swift:
```diff
@@ -0,0 +1,170 @@
+    @objc @Published public var backgroundDimmed: Bool = false
+}
+
+public typealias DrawerStateChangedCompletionBlock = (_ isExpanded: Bool) -> Void
```
Calling this a completion block seems very "objC"ish...
I vote for just deleting this typealias and declaring line 22 as follows:

` public var onStateChange: (() -> Void)?`

rdeassis commented on ios/FluentUI/DrawerVnext/Drawer.swift:
```diff
@@ -0,0 +1,170 @@
+    @ObservedObject public var tokens = DrawerTokens()
+
+    // keep track of dragged offset
+    @State internal var draggedOffsetWidth: CGFloat?
```
rename to dragOffset?

kubalani replied:
how about `dragOffsetX`? I need some context to indicate this is along x-axis only as it's a float and not a size

rdeassis replied:
how about horizontaldragOffset?

rdeassis commented on ios/FluentUI/DrawerVnext/Drawer.swift:
```diff
@@ -0,0 +1,170 @@
+    @State internal var draggedOffsetWidth: CGFloat?
+
+    // internal drawer state
+    @State internal var isDrawerOpen: Bool = false
```
How about renaming to **isContentPresented**?

kubalani replied:
I like this name, much better.

rdeassis commented on ios/FluentUI/DrawerVnext/Drawer.swift:
```diff
@@ -0,0 +1,170 @@
+    }
+
+    private var backgroundLayerOpacity: Double {
+        return state.backgroundDimmed ? 0.5 : 0
```
This opacity value should be tokenized. Can you start a conversation with Victor about it? (and open a GitHub issue for it)

rdeassis commented on ios/FluentUI/DrawerVnext/Drawer.swift:
```diff
@@ -0,0 +1,170 @@
+// MARK: - Utility + Helpers
+
+extension GeometryProxy {
+    func portraitOrientationAgnosticSize() -> CGSize {
```
Can you move this into the implementation of the drawer as a private func that takes the GeometryProxy as a parameter? This function does not make sense for the GeometryProxy out of the Drawer context and that's why I don't think it should be an extension.

kubalani replied:
Extension methods are only a (pretty nice) syntactic sugar for a static helper method. I prefer extensions methods because they are much easier to read as compare to private helper methods. We talk about this more offline.

rdeassis replied:
We can discuss it offline. I believe the same argument applies here: we need to make sure that the extension makes sense out of the Drawer context, or at least make it private so it's not visible everywhere in the library where it doesn't make sense.

kubalani replied:
I have removed the extension and converted to a private func, we can revisit this when we add more functionality to drawer

rdeassis commented on ios/FluentUI/DrawerVnext/Drawer.swift:
```diff
@@ -0,0 +1,170 @@
+extension GeometryProxy {
+    func portraitOrientationAgnosticSize() -> CGSize {
+        let isPortraitMode = self.size.width < self.size.height
+        if isPortraitMode {
```
nit: put the condition directly in the if statement?

kubalani replied:
that would make it hard for readability. Which one of this you would prefer?

```swift
let isPortraitMode = self.size.width < self.size.height
        if isPortraitMode {
            return CGSize(width: self.size.width, height: self.size.height)
        } else {
            return CGSize(width: self.size.height, height: self.size.width)
        }
```

```swift
        // check if user is in portrait mode
        if self.size.width < self.size.height {
            return CGSize(width: self.size.width, height: self.size.height)
        } else {
            return CGSize(width: self.size.height, height: self.size.width)
        }
```

rdeassis replied:
I guess the second option is what I tried to suggest (putting the condition you're storing in isPortraitMode directly in the if statement).

rdeassis commented on ios/FluentUI/DrawerVnext/Drawer.swift:
```diff
@@ -0,0 +1,170 @@
+// MARK: - Utility + Helpers
+
+extension GeometryProxy {
+    func portraitOrientationAgnosticSize() -> CGSize {
```
How about calling the method **sizeForCurrentOrientation**?

kubalani replied:
`sizeForCurrentOritentation` seems like to expect a parameter i.e current orientation. The function needs to return portrait orientation size regardless of current orientation. How about `portraitOrientationSize`?

rdeassis replied:
portraitOrientationSize sounds like it does not return a value if it's in landscape.
How about **sizeInCurrentOrientation** or **sizeInActualOrientation**?

rdeassis commented on ios/FluentUI/DrawerVnext/Drawer.swift:
```diff
@@ -0,0 +1,170 @@
+
+// MARK: - Previews
+
+import SwiftUI
```
remove duplicate import

rdeassis commented on ios/FluentUI/DrawerVnext/SlideOverPanel.swift:
```diff
@@ -0,0 +1,174 @@
+
+    private func resolvedContentOffset() -> CGFloat {
+        // override offset if required
+        if let preferredContentWidth = preferredContentOffset {  // parent view wants to take over primarily to conform to user drag gesture
```
nit: comment on the next line

rdeassis commented on ios/FluentUI/DrawerVnext/SlideOverPanel.swift:
```diff
@@ -0,0 +1,174 @@
+public struct InteractiveSpacer: View {
+    public var body: some View {
+        ZStack {
+            Color.black.opacity(0.001)
```
I think the same applies here for a tokenized opacity that can be tracked in the same issue I mentioned before.
Could you talk to Victor about it?

kubalani replied:
Victor is OOF in the meanwhile I have added a transparent color to stylesheet. I will discuss this with him as a part of #358

rdeassis commented on ios/FluentUI/DrawerVnext/Drawer.swift:
```diff
@@ -0,0 +1,170 @@
+                .direction(slideOutDirection)
+                .width(proxy.portraitOrientationAgnosticSize().width)
+                .performOnBackgroundTap {
+                    state.isExpanded.toggle()
```
Is there a scenario where tapping on the background will expand the drawer?
I believe there's not and this should be just **state.isExpanded = false**

kubalani commented on ios/FluentUI.Demo/FluentUI.Demo/Demos/DrawerVnextDemoController.swift:
```diff
@@ -0,0 +1,174 @@
+    }
+}
+
+extension UIStackView {
```
The scope of extension is internal by default. If this was intended to be used as a multiple times I would have used a broader modifier (`public` or `open`) 

The only reason an extension is used here is for code separation. It makes code readability much easier
```swift
 addBackground(stackview: container, color: Colors.surfacePrimary)
```

vs 

```swift
 container.addBackground(color: Colors.surfacePrimary)
```

I would be interested to know if there are any specific reason we are trying to limit extension or having excessive extensions is bad for the code?

rdeassis replied:
The helper function is specific for this scenario and not for all uses of the UIStackView.
You don't want to have this addBackground available for UIStackView instances everywhere if it only makes sense here. It makes the code more complicated and hard to maintain if this gets used somewhere else and you also need to change this here.

You could argue that making this extension private could solve that, but this is being used only once for a very specific scenario... which makes the extension unnecessary.

kubalani commented on ios/FluentUI.Demo/FluentUI.Demo/Demos/DrawerVnextDemoController.swift:
```diff
@@ -0,0 +1,174 @@
+        for view in actionViews() {
+            container.addArrangedSubview(view)
+        }
+        container.addBackground(color: Colors.surfacePrimary)
```
This is demo code not `FluentUI`. Since we are sharing components like images, Democontroller with rest of the Demo code the Colors library is part of it. Adding `StyleSheetManager` to demo code would give an indication that it's ok for client to use the internal style or theme sheet for their own content. 

We don't want client to use `StyleSheetManager` directly. At some point we will replace `Color` wrapper to use `Stylesheet` and clients will continue using their current API. At least that is what I think is the expected functionality.

rdeassis replied:
The client app will use StylesheetManager directly to override Themes in our current design. That is where it can access all the appearance proxies (including colors).

Although it's not a big deal to have these calls here now, it'll be more work when we discontinue the Colors class.

kubalani commented on ios/FluentUI.Demo/FluentUI.Demo/Demos/DrawerVnextDemoController.swift:
```diff
@@ -0,0 +1,174 @@
+        }
+    }
+
+    @objc func drawerDidChangeState(state: DrawerState, controller: UIViewController) {
```
This is intended for the client to show how does the delegates work. It would encourage them to use this rather than manipulating state. 

Is there any thing wrong with keeping it?

rdeassis replied:
I don't think it's necessarily wrong to keep it, but this is an optional method (which wouldn't help us catch func signature changes, right?).
We usually avoid doing just print() since it's not necessarily visible in the demo app.
You can keep it if you're very passionate about it.

kubalani replied:
I have removed it for now, since empty function might look like a bug. We can add this as an example later while adding a mock teams demo controller

kubalani commented on ios/FluentUI/DrawerVnext/Drawer+Drag.swift:
```diff
@@ -0,0 +1,31 @@
+
+import SwiftUI
+
+internal extension Drawer {
```
This will add a significant code when while addressing #356 . SwiftUI encourage modular structure as opposed monolithic views.

rdeassis replied:
You're not making it modular if your extension is specific to just the Drawer, are you?

Like I said before: if this will add significant amounts of code when you implement the gesture presentation, I can see the value for it.

To me this seems not needed for now.

kubalani replied:
I have converted this to a private func, I will defer the need for gesture extension and add when required.

kubalani commented on ios/FluentUI/DrawerVnext/Drawer+Drag.swift:
```diff
@@ -0,0 +1,31 @@
+            .onChanged { value in
+                let withinDragBounds = state.presentationDirection == .left ? value.translation.width < 0 : value.translation.width > 0
+                if withinDragBounds {
+                    draggedOffsetWidth = value.translation.width
```
dragOffset is a float and can be across horizontal axis or vertical axis. May be `dragOffsetX`?

rdeassis replied:
horizontalDragOffiset?

kubalani replied:
yup, this is a much better name. I have updated it !!

rdeassis commented on ios/FluentUI/DrawerVnext/SlideOverPanel.swift:
```diff
@@ -0,0 +1,174 @@
+    public var body: some View {
+        ZStack {
+            defaultBackgroundColor
+        }
```
.contentShape(Rectangle())

Adding this modifier here will enable hit testing even if the color is transparent.
This should fix the "tap out of the drawer dismiss bug".

kubalani replied:
this worked, updated !!

rdeassis commented on ios/FluentUI/DrawerVnext/SlideOverPanel.swift:
```diff
@@ -0,0 +1,174 @@
+        HStack {
+            if direction == .right {
+                InteractiveSpacer(defaultBackgroundColor: $tokens.backgroundClearColor)
+                    .onTapGesture (count: 1, perform: actionOnBackgroundTap ?? {})
```
you don't need the count:1 parameter here, do you? (same in line 62)

rdeassis reviewed (approved):
Thanks for addressing all the comments!
I left a few more (including a bug on the "tap outside to dismiss"), but the .contentShape(Rectangle()) modifier will fix it.

kubalani commented on ios/FluentUI/DrawerVnext/Drawer.swift:
```diff
@@ -0,0 +1,209 @@
+            onStateChange?()
+        }
+    }
+    @objc @Published public var presentationDirection: DrawerDirection = .left
```
Discuss this offline with Rafael, published property is not required for direction and animation duration

kubalani commented on ios/FluentUI/DrawerVnext/SideOverPanel+Modifiers.swift:
```diff
@@ -0,0 +1,78 @@
+    /// Add action or callback to be executed when background view is Tapped
+    /// - Parameter `performOnBackgroundTap`:  defaults to no-op
+    /// - Returns: `SlideOverPanel`
+    func performOnBackgroundTap(_ performOnBackgroundTap: (() -> Void)?) -> SlideOverPanel {
```
this is now fixed. thanks for catching this. I will add a new issues to add integration tests

kubalani commented:
> Thanks for addressing all the comments!
> I left a few more (including a bug on the "tap outside to dismiss"), but the .contentShape(Rectangle()) modifier will fix it.

Thanks for reviews
